drivers/can/mcp2515.c

```diff
@@ -0,0 +1,618 @@
+
+	const u8_t cnf2 = btlmode | sam | phseg1 | prseg;
+
+	/* CNF3; SOF<7>|WAKFIL<6>|UND<5:3>|PHSEG2<2:0> */
+	const u8_t sof = 0 << 7;
+	const u8_t wakfil = 0 << 6;
+	const u8_t und = 0 << 3;
+	const u8_t phseg2 = (dev_cfg->tq_bs2 - 1);
+
+	const u8_t cnf3 = sof | wakfil | und | phseg2;
+
+	/* CANINTE
+	 * MERRE<7>:WAKIE<6>:ERRIE<5>:TX2IE<4>:TX1IE<3>:TX0IE<2>:RX1IE<1>:RX0IE<0>
+	 * all TX and RX buffer interrupts enabled
+	 */
+	const u8_t caninte = BIT(4) | BIT(3) | BIT(2) | BIT(1) | BIT(0);
+
+	/* Receive everything, filtering done in driver, RXB0 roll over into RXB1 */
+	const u8_t rx0_ctrl = BIT(6) | BIT(5) | BIT(2);
+	const u8_t rx1_ctrl = BIT(6) | BIT(5);
+
+	__ASSERT((cfg->tq_sjw >= 1) && (cfg->tq_sjw <= 4), "1 <= SJW <= 4");
+	__ASSERT((cfg->tq_prop >= 1) && (cfg->tq_prop <= 8), "1 <= PROP <= 8");
+	__ASSERT((cfg->tq_bs1 >= 1) && (cfg->tq_bs1 <= 8), "1 <= BS1 <= 8");
+	__ASSERT((cfg->tq_bs2 >= 2) && (cfg->tq_bs2 <= 8), "2 <= BS2 <= 8");
+	__ASSERT(cfg->tq_prop + cfg->tq_bs1 >= cfg->tq_bs2, "PROP + BS1 >= BS2");
+	__ASSERT(cfg->tq_bs2 > cfg->tq_sjw, "BS2 > SJW");
+
+	if (CONFIG_CAN_MCP2515_OSC_FREQ % (bit_length * bitrate * 2)) {
+		SYS_LOG_ERR("Prescaler is not a natural number! "
+			    "prescaler = osc_rate / ((PROP + SEG1 + SEG2 + 1) * bitrate * 2)"
+			    "prescaler = %d / ((%d + %d + %d + 1) * %d * 2)",
+			    CONFIG_CAN_MCP2515_OSC_FREQ,
+			    dev_cfg->tq_prop,
+			    dev_cfg->tq_bs1,
+			    dev_cfg->tq_bs2,
+			    bitrate);
+	}
+
+	config_buf[0] = cnf3;
+	config_buf[1] = cnf2;
+	config_buf[2] = cnf1;
+	config_buf[3] = caninte;
+
+	/* will enter configuration mode automatically */
+	mcp2515_cmd_soft_reset(dev);
+
+	mcp2515_cmd_write_reg(dev, MCP2515_ADDR_CNF3, config_buf, 4);
+
+	mcp2515_cmd_bit_modify(dev, MCP2515_ADDR_RXB0CTRL, rx0_ctrl, rx0_ctrl);
+	mcp2515_cmd_bit_modify(dev, MCP2515_ADDR_RXB1CTRL, rx1_ctrl, rx1_ctrl);
+
+	return mcp2515_set_mode(dev,
+				mcp2515_convert_can_mode_to_mcp2515_mode(mode));
+}
+
+int mcp2515_send(struct device *dev, struct can_msg *msg, s32_t timeout,
+		 can_tx_callback_t callback)
+{
+	struct mcp2515_data *dev_data = DEV_DATA(dev);
+	int tx_idx = 0;
+	u8_t addr_tx_ctrl;
+	u8_t tx_frame[MCP2515_FRAME_LEN];
+
+	if (k_sem_take(&dev_data->tx_sem, timeout) != 0) {
+		return CAN_TIMEOUT;
+	}
+
+	k_mutex_lock(&dev_data->tx_mutex, K_FOREVER);
+
+	/* find a free tx slot */
+	for (; tx_idx < MCP2515_TX_CNT; tx_idx++) {
+		if ((BIT(tx_idx) & dev_data->tx_busy_map) == 0) {
+			dev_data->tx_busy_map |= BIT(tx_idx);
+			break;
+		}
+	}
+
+	k_mutex_unlock(&dev_data->tx_mutex);
+
+	if (tx_idx == MCP2515_TX_CNT) {
+		SYS_LOG_WRN("no free tx slot available");
+		return CAN_TX_ERR;
+	}
+
+	dev_data->tx_cb[tx_idx].cb = callback;
+
+	addr_tx_ctrl = MCP2515_ADDR_TXB0CTRL + (tx_idx *
+			MCP2515_ADDR_OFFSET_FRAME2FRAME);
+
+	mcp2515_convert_can_msg_to_mcp2515_frame(msg, tx_frame);
+	mcp2515_cmd_write_reg(dev, addr_tx_ctrl + MCP2515_ADDR_OFFSET_CTRL2FRAME,
+			tx_frame, MCP2515_FRAME_LEN);
```

Comment:
tiny alignement issue (should be below last '(' + 1 space)

apply that rule everywhere (I probably did not notice it the first time)

Reply:
Sorry I have to double check here, as the first time I did it checkpatch complained about space instead of tabs
You want any link breaks aligned under and after the last current open bracket right? It doesn't have to be exactly 1 behind the last open bracket, right? Anyways, I did just that and checkpatch didn't complain (but keeps seeing lines longer than 80 chars that aren't, does it tab extend to 8 space?)